File: src/SampleCRM.Web/Services/CustomersService.cs
using OpenRiaServices.DomainServices.Hosting;
using OpenRiaServices.DomainServices.Server;
using SampleCRM.Web.Attributes;
using SampleCRM.Web.Models;
using System;
using System.Data.Entity.Migrations;
using System.Linq;

namespace SampleCRM.Web
{
    [EnableClientAccess]
    public class CustomersService : SampleCRMService
    {
        [Query]
        public IQueryable<Customers> GetCustomers(string search) =>
            _context.Customers
                    .Where(x => x.FirstName.ToLower().Contains(search.ToLower())
                            || x.LastName.ToLower().Contains(search.ToLower())
                            || search == "")
                    .OrderBy(c => c.FirstName)
                    .ThenBy(c => c.LastName);

        [Query]
        public IQueryable<Customers> GetCustomersCombo() =>
            GetCustomers(string.Empty);

        [Query]
        public IQueryable<Customers> GetCustomersWithoutPictures(string search) =>
            GetCustomers(search)
            .ToList()
            .Select(x => new Customers
            {
                CustomerID = x.CustomerID,
                AddressLine1 = x.AddressLine1,
                AddressLine2 = x.AddressLine2,
                BirthDateUTC = x.BirthDateUTC,
                ChildrenAtHome = x.ChildrenAtHome,
                City = x.City,
                CountryCode = x.CountryCode,
                CreatedOnUTC = x.CreatedOnUTC,
                Education = x.Education,
                EmailAddress = x.EmailAddress,
                FirstName = x.FirstName,
                Phone = x.Phone,
                NumberCarsOwned = x.NumberCarsOwned,
                Gender = x.Gender,
                YearlyIncome = x.YearlyIncome,
                Suffix = x.Suffix,
                IsHouseOwner = x.IsHouseOwner,
                LastName = x.LastName,
                MaritalStatus = x.MaritalStatus,
                MiddleName = x.MiddleName,
                Occupation = x.Occupation,
                PostalCode = x.PostalCode,
                LastModifiedOnUTC = x.LastModifiedOnUTC,
                Region = x.Region,
                SearchTerms = x.SearchTerms,
                Title = x.Title,
                TotalChildren = x.TotalChildren,
                Picture = new byte[] { 0 }
            }).AsQueryable();

        [Query]
        public IQueryable<Customers> GetLatestCustomers(string search, int limit) =>
            GetCustomersWithoutPictures(search)
                .OrderByDescending(x => x.CreatedOnUTC)
                .Take(limit);

        [Delete]
        [RestrictAccessReadonlyMode]
        public void DeleteCustomer(Customers customer)
        {
            var dCustomer = _context.Customers.FirstOrDefault(x => x.CustomerID == customer.CustomerID);
            if (dCustomer == null)
                return;

            _context.Customers.Remove(dCustomer);
            _context.SaveChanges();
        }

        [Insert]
        [RestrictAccessReadonlyMode]
        public void InsertCustomer(Customers customer)
        {
            customer.CustomerID = new Random().Next((int)Math.Pow(10, 12), (int)Math.Pow(10, 13) - 1);
            if (customer.CustomerID < 0)
                customer.CustomerID *= -1;

            customer.LastModifiedOnUTC = customer.CreatedOnUTC = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _context.Customers.Add(customer);
            _context.SaveChanges();
        }

        [Update]
        [RestrictAccessReadonlyMode]
        public void UpdateCustomer(Customers customer)
        {
            customer.LastModifiedOnUTC = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _context.Customers.AddOrUpdate(customer);
            _context.SaveChanges();
        }

        public Customers GetCustomerById(long customerId) =>
            GetCustomers(string.Empty).SingleOrDefault(x => x.CustomerID == customerId);

        public byte[] GetCustomerPicture(long customerId)
        {
            var customer = GetCustomerById(customerId);
            return customer != null ? customer.Picture : throw new ArgumentNullException($"No Such Customer {customerId}");
        }
    }
}

File: src/SampleCRM.Web/Services/CustomersService.vb
Imports OpenRiaServices.DomainServices.Hosting
Imports OpenRiaServices.DomainServices.Server
Imports SampleCRM.Web.Attributes
Imports System
Imports System.Data.Entity.Migrations
Imports System.Linq

Namespace SampleCRM.Web
    <EnableClientAccess>
    Public Class CustomersService
        Inherits SampleCRMService

        <Query>
        Public Function GetCustomers() As IQueryable(Of Customers)
            Return _context.Customers
        End Function

        <Query>
        Public Function GetLatestCustomers(ByVal limit As Integer) As IQueryable(Of Customers)
            Return _context.Customers.OrderByDescending(Function(x) x.CreatedOn).Take(limit)
        End Function

        <Delete>
        <RestrictAccessDeveloperMode>
        Public Sub DeleteCustomer(ByVal customer As Customers)
            Dim dCustomer = _context.Customers.FirstOrDefault(Function(x) x.CustomerID = customer.CustomerID)
            If dCustomer Is Nothing Then Return
            _context.Customers.Remove(dCustomer)
            _context.SaveChanges()
        End Sub

        <Insert>
        <RestrictAccessDeveloperMode>
        Public Sub InsertCustomer(ByVal customer As Customers)
            customer.CustomerID = New Random().[Next](CInt(Math.Pow(10, 12)), CInt(Math.Pow(10, 13)) - 1)
            If customer.CustomerID < 0 Then customer.CustomerID = customer.CustomerID * -1
            customer.CreatedOn = Date.Now.ToString()
            customer.LastModifiedOn = customer.CreatedOn
            _context.Customers.Add(customer)
            _context.SaveChanges()
        End Sub

        <Update>
        <RestrictAccessDeveloperMode>
        Public Sub UpdateCustomer(ByVal customer As Customers)
            customer.LastModifiedOn = Date.Now.ToString()
            _context.Customers.AddOrUpdate(customer)
            _context.SaveChanges()
        End Sub

        Private Class CSharpImpl
            <Obsolete("Please refactor calling code to use normal Visual Basic assignment")>
            Shared Function __Assign(Of T)(ByRef target As T, value As T) As T
                target = value
                Return value
            End Function
        End Class
    End Class
End Namespace
